extensions/amp-bind/0.1/bind-validator.js

```diff
@@ -0,0 +1,334 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @typedef {{
+ *   allowedProtocols: (!Object<string,boolean>|undefined),
+ *   blockedURLs: (Array<string>|undefined),
+ * }}
+ */
+let PropertyRulesDef;
+
+/**
+ * Property rules that apply to any and all tags.
+ * @private {Object<string, ?PropertyRulesDef>}
+ */
+const GLOBAL_PROPERTY_RULES = {
+  'class': {
+    blacklistedValueRegex: '(^|\\W)i-amphtml-',
+  },
+};
+
+/**
+ * Maps tag names to property names to PropertyRulesDef.
+ * If `ELEMENT_RULES[tag][property]` is null, then all values are valid
+ * for that property in that tag.
+ * @private {Object<string, Object<string, ?PropertyRulesDef>>}}
+ */
+const ELEMENT_RULES = createElementRules_();
+
+/**
+ * BindValidator performs runtime validation of Bind expression results.
+ *
+ * For performance reasons, the validation rules enforced are a subset
+ * of the AMP validator's, selected with a focus on security and UX.
+ */
+export class BindValidator {
+  /**
+   * Returns true if (tag, property) binding is allowed.
+   * Otherwise, returns false.
+   * @note `tag` and `property` are case-sensitive.
+   * @param {!string} tag
+   * @param {!string} property
+   * @return {boolean}
+   */
+  canBind(tag, property) {
+    return (this.rulesForTagAndProperty_(tag, property) !== undefined);
+  }
+
+  /**
+   * Returns true if `value` is a valid result for a (tag, property) binding.
+   * Otherwise, returns false.
+   * @param {!string} tag
+   * @param {!string} property
+   * @param {?string} value
+   * @return {boolean}
+   */
+  isResultValid(tag, property, value) {
+    const attrRules = this.rulesForTagAndProperty_(tag, property);
+
+    // If binding to (tag, property) is not allowed, return false.
+    if (attrRules === undefined) {
+      return false;
+    }
+
+    // If binding is allowed but have no specific rules, return true.
+    if (attrRules === null) {
+      return true;
+    }
+
+    // @see validator/engine/validator.ParsedUrlSpec.validateUrlAndProtocol()
+    const allowedProtocols = attrRules.allowedProtocols;
+    if (allowedProtocols && value) {
+      const re = /^([^:\/?#.]+):.*$/;
```

Comment:
`.*` will not match new line which allows XSS in the form of `javascript:alert("XSS")\n` as it won't get a match and will skip the protocol validation. (`anchor.setAttribute('href', 'javascript:alert("XSS")\n');` executes perfectly fine (https://jsfiddle.net/fmuwsgd7/)